functions/airtable.js

```diff
@@ -244,37 +294,265 @@ async function getBulkOrder(records) {
+   */
+  constructor(
+    intakeRecord,
+    requested,
+    provided,
+    bulkDeliveryRoute,
+    volunteer,
+    itemToCategory,
+  ) {
+    this.intakeRecord = intakeRecord;
+    this.requested = requested;
+    this.provided = provided;
+    this.bulkDeliveryRoute = bulkDeliveryRoute;
+    this.volunteer = volunteer;
+    this.itemToCategory = itemToCategory;
+  }
+
+  bulkPurchasedItemsByGroup() {
+    return _.groupBy(_.toPairs(this.provided), ([item]) => {
+      return _.includes(['Bread', 'Bananas'], item)
+        ? 'Last'
+        : this.itemToCategory[item];
+    });
+  }
+
+  /**
+   * @returns {[{ item: string, quantity: number | null }]} List of items and quantities.
+   */
+  getAdditionalItems() {
+    const fields = this.intakeRecord[1];
+    const missingItems = _.filter(
+      _.map(_.toPairs(this.requested), ([item, numRequested]) => {
+        return { item, quantity: numRequested - _.get(this.provided, item, 0) };
+      }),
+      ({ quantity }) => {
+        return quantity !== 0;
+      }
+    );
+
+    const customItems = !_.isNull(fields.otherItems)
+      ? _.map(
+        _.filter(
+          _.map(fields.otherItems.split(','), (item) => {
+            return item.trim();
+          }),
+          (item) => {
+            return item.length > 0;
+          }
+        ),
+        (item) => {
+          return { item, quantity: null };
+        }
+      )
+      : [];
+
+    // @ts-ignore eslint doesn't understand array structures I guess?
+    return missingItems.concat(customItems);
+  }
+}
+
+/**
+ * Reconcile bulk delivery orders with procured items from Bulk Order table.
+ * @param {[string, Object, Object][]} allRoutes All bulk delivery routes for this week.
+ * @param {Date} deliveryDate Date these orders will go out.
+ * @returns {Promise<ReconciledOrder[]>} List of reconciled orders.
+ */
+async function reconcileOrders(allRoutes, deliveryDate) {
+  const itemToCategory = _.fromPairs(
+    _.map(await getAllRecords(ITEMS_BY_HOUSEHOLD_SIZE_TABLE), ([, fields]) => {
+      return [fields.item, fields.category];
+    })
+  );
+
+  const intakeRecords = await getTicketsForRoutes(allRoutes);
+
+  const itemToNumAvailable = await getItemToNumAvailable(deliveryDate);
+
+  const volunteerRecords = await getAllRecords(VOLUNTEER_FORM_TABLE);
+
+  const routesByKey = _.fromPairs(
+    _.map(allRoutes, ([id, fields,]) => {
+      return [id, fields];
+    })
+  );
+
+  const volunteersByKey = _.fromPairs(
+    _.map(volunteerRecords, ([id, fields,]) => {
+      return [id, fields];
+    })
+  );
+
+  // WARNING: We cannot process orders in an async manner. We need them to be
+  // processed in a deterministic order, so that the accounting (which tickets
+  // need custom shoppers to buy bulk items we couldn't procure) will be the
+  // same between packing slips and shopping lists. Therefore, we need to be
+  // careful that we sort tickets the same way every time, and we do accounting
+  // for them sequentially.
```

Comment:
I got the point from your comment; it took me a minute to process tho. An example could help make this a bit more comprehensible.

Reply:
Added one!